mlflow/store/tracking/file_store.py

```diff
@@ -1319,3 +1326,372 @@ def _read_helper(root, file_name, attempts_remaining=2):
+
+        Returns:
+            The updated TraceInfo object.
+        """
+        trace_info, trace_dir = self._get_trace_info_and_dir(request_id)
+        trace_info.execution_time_ms = timestamp_ms - trace_info.timestamp_ms
+        trace_info.status = status
+        trace_info.request_metadata.update(request_metadata)
+        trace_info.tags.update(tags)
+        self._save_trace_info(trace_info, trace_dir, overwrite=True)
+        return trace_info
+
+    def get_trace_info(self, request_id: str) -> TraceInfo:
+        """
+        Get the trace matching the `request_id`.
+
+        Args:
+            request_id: String id of the trace to fetch.
+
+        Returns:
+            The fetched Trace object, of type ``mlflow.entities.TraceInfo``.
+        """
+        return self._get_trace_info_and_dir(request_id)[0]
+
+    def _get_trace_info_and_dir(self, request_id: str) -> Tuple[TraceInfo, str]:
+        trace_dir = self._find_trace_dir(request_id, assert_exists=True)
+        trace_info = self._get_trace_info_from_dir(trace_dir)
+        if trace_info.request_id != request_id:
+            raise MlflowException(
+                f"Trace with request ID '{request_id}' metadata is in invalid state.",
+                databricks_pb2.INVALID_STATE,
+            )
+        return trace_info, trace_dir
+
+    def _find_trace_dir(self, request_id, assert_exists=False):
+        self._check_root_dir()
+        all_experiments = self._get_active_experiments(True) + self._get_deleted_experiments(True)
+        for experiment_dir in all_experiments:
+            traces_dir = os.path.join(experiment_dir, FileStore.TRACES_FOLDER_NAME)
+            if exists(traces_dir):
+                traces = find(traces_dir, request_id, full_path=True)
+                if len(traces) == 0:
+                    continue
+                return traces[0]
+        if assert_exists:
+            raise MlflowException(
+                f"Trace with request ID '{request_id}' not found",
+                RESOURCE_DOES_NOT_EXIST,
+            )
+
+    def _get_trace_info_from_dir(self, trace_dir) -> TraceInfo:
+        meta = FileStore._read_yaml(trace_dir, FileStore.META_DATA_FILE_NAME)
+        trace_info = self._convert_dict_to_trace_info(meta)
+        trace_info.request_metadata = self._get_dict_from_trace_sub_folder(
+            trace_dir, FileStore.TRACE_REQUEST_METADATA_FOLDER_NAME
+        )
+        trace_info.tags = self._get_dict_from_trace_sub_folder(
+            trace_dir, FileStore.TRACE_TAGS_FOLDER_NAME
+        )
+        return trace_info
+
+    def set_trace_tag(self, request_id: str, key: str, value: str):
+        """
+        Set a tag on the trace with the given request_id.
+
+        Args:
+            request_id: The ID of the trace.
+            key: The string key of the tag.
+            value: The string value of the tag.
+        """
+        trace_info, trace_dir = self._get_trace_info_and_dir(request_id)
```

Comment:
As we store tags as separate file, can we avoid reading and merging the entire info and just edit the trace file only? Mainly for efficiency, but also to avoid accidental update on immutable objects like trace info and request metadata🙂

Reply:
Yep make sense